metadata/reader_test.go

```diff
@@ -33,47 +33,260 @@
+				testutil.Dir("bar/"),
+				testutil.Link("bar/foolink", "foo"),
+				testutil.Link("bar/foolink2", "bar/foolink"),
+				testutil.Dir("bar/1/"),
+				testutil.File("bar/1/baz.txt", "testtest"),
+				testutil.Link("barlink", "bar/1/baz.txt"),
+				testutil.Symlink("foosym", "bar/foolink2"),
+			},
+			want: []check{
+				numOfNodes(6), // root dir + 2 dirs + 1 flie(linked) + 1 file(linked) + 1 symlink
+				hasFile("foo", 6),
+				hasOwner("foo", 1000, 1000),
+				hasFile("bar/foolink", 6),
+				hasOwner("bar/foolink", 1000, 1000),
+				hasFile("bar/foolink2", 6),
+				hasOwner("bar/foolink2", 1000, 1000),
+				hasFile("bar/1/baz.txt", 8),
+				hasFile("barlink", 8),
+				hasDirChildren("bar", "foolink", "foolink2", "1"),
+				hasDirChildren("bar/1", "baz.txt"),
+				sameNodes("foo", "bar/foolink", "bar/foolink2"),
+				sameNodes("bar/1/baz.txt", "barlink"),
+				linkName("foosym", "bar/foolink2"),
+				hasNumLink("foo", 3),     // parent dir + 2 links
+				hasNumLink("barlink", 2), // parent dir + 1 link
+				hasNumLink("bar", 3),     // parent + "." + child's ".."
+			},
+		},
+		{
+			name: "various files",
+			in: []testutil.TarEntry{
+				testutil.Dir("bar/"),
+				testutil.File("bar/../bar///////////////////foo", ""),
+				testutil.Chardev("bar/cdev", 10, 11),
+				testutil.Blockdev("bar/bdev", 100, 101),
+				testutil.Fifo("bar/fifo"),
+			},
+			want: []check{
+				numOfNodes(6), // root dir + 1 file + 1 dir + 1 cdev + 1 bdev + 1 fifo
+				hasFile("bar/foo", 0),
+				hasChardev("bar/cdev", 10, 11),
+				hasBlockdev("bar/bdev", 100, 101),
+				hasFifo("bar/fifo"),
+			},
+		},
+	}
+	for _, tt := range tests {
+		for _, prefix := range allowedPrefix {
+			prefix := prefix
+			for srcCompresionName, srcCompression := range srcCompressions {
+				t.Run(tt.name+"-"+srcCompresionName, func(t *testing.T) {
+					opts := []testutil.BuildTarOption{
+						testutil.WithPrefix(prefix),
+					}
+
+					ztoc, sr, err := ztoc.BuildZtocReader(t, tt.in, srcCompression, 64, opts...)
+					if err != nil {
+						t.Fatalf("failed to build ztoc: %v", err)
+					}
+					telemetry, checkCalled := newCalledTelemetry()
+
+					// create a metadata reader
+					r, err := newTestableReader(sr, ztoc.TOC, WithTelemetry(telemetry))
+					if err != nil {
+						t.Fatalf("failed to create new reader: %v", err)
+					}
+					defer r.Close()
+					t.Logf("vvvvv Node tree vvvvv")
+					t.Logf("[%d] ROOT", r.RootID())
+					dumpNodes(t, r, r.RootID(), 1)
+					t.Logf("^^^^^^^^^^^^^^^^^^^^^")
+					for _, want := range tt.want {
+						want(t, r)
+					}
+					if err := checkCalled(); err != nil {
+						t.Errorf("telemetry failure: %v", err)
+					}
+				})
+			}
+		}
+	}
 }
 
-func newTestableReader(sr *io.SectionReader, toc ztoc.TOC, opts ...Option) (testableReader, error) {
-	f, err := os.CreateTemp("", "readertestdb")
-	if err != nil {
-		return nil, err
+func BenchmarkMetadataReader(b *testing.B) {
```

Comment:
Does this ever get called?

Reply:
No, this is a benchmark function (similar to a TestXXX), so it will never get invoked outside of `go test -run=TestXXX` or `go test -bench=BenchmarkXXX`.